include/mbedtls/ssl.h

```diff
@@ -3851,25 +3851,35 @@ int mbedtls_ssl_renegotiate( mbedtls_ssl_context *ssl );
 int mbedtls_ssl_read( mbedtls_ssl_context *ssl, unsigned char *buf, size_t len );
 
 /**
- * \brief          Try to write exactly 'len' application data bytes
+ * \brief          Queue application data for dispatch
  *
- * \warning        This function will do partial writes in some cases. If the
- *                 return value is non-negative but less than length, the
- *                 function must be called again with updated arguments:
- *                 buf + ret, len - ret (if ret is the return value) until
- *                 it returns a value equal to the last 'len' argument.
+ * \warning        This function may queue less data for dispatch than requested.
+ *                 If the return value is non-negative but less than \p len,
+ *                 the function should be called again for the remaining data.
  *
- * \param ssl      SSL context
- * \param buf      buffer holding the data
- * \param len      how many bytes must be written
+ * \warning        This function does not send any of the application data
+ *                 provided by the user, but only prepares and queues the
+ *                 corresponding protected application data records. Dispatch to
+ *                 the underlying transport is only attempted when the user
+ *                 explicitly calls mbedtls_ssl_flush(), or when a subsequent
+ *                 call to mbedtls_ssl_write() can not make progress without
+ *                 dispatching some queued application records first.
```

Comment:
This seems contradictory with “This function does not send any of the application data” above. Can `mbedtls_ssl_write` cause application data to be written or not?

(Generally https://github.com/ARMmbed/mbedtls/issues/4183 is a lot clearer than the documentation here.)

Reply:
Good point. `mbedtls_ssl_write()` may try to deliver application data from previous calls to `mbedtls_ssl_write()`, but it will _never_ attempt to deliver application data from the current call. This way, you always know what a WANT_READ/WRITE referred to previous calls, not the current one, and that the current one therefore has to be retried.

Reply:
I will clarify the documentation.